### PR TITLE
Fix comp

### PR DESCRIPTION
**Pull request by tmontaigu**
closes: _please link all relevant issues_

### Check-list:

* [ ] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] Relevant issues are marked as resolved/closed, related issues are linked in the description
* [ ] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

**Comment by IceTDrinker** [comment]
does this need a back port ?

**Comment by IceTDrinker** [comment]
i.e. is it a noise issue or a correctness issue separate from the noise concern ?

**Comment by tmontaigu** [comment]
It fixes a theoretical noise issue (it never failed in 1 year of tests)

File: tfhe/src/integer/server_key/comparator.rs [hunk 1]
```diff
@@ -21,6 +21,24 @@ enum MinMaxSelector {
...
+/// This function encodes how to reduce the ordering of two blocks.
+///
+/// It is used to generate the necessary lookup table.
+///
+/// `x` is actually two ordering values packed
+/// where in the first 2 lsb bits there is the ordering value of the less significant block,
+/// and the 2 msb bits of x contains the ordering of the more significant block.
+fn reduce_two_orderings_function(x: u64) -> u64 {
+    let msb = (x >> 2) & 3;
+    let lsb = x & 3;
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/comparator.rs:33]: can we do better ? i.e. do we need two full bits for each packed ordering ?

From afar for an ordering you need lower, equal, greater or 3 values, which could be packed in a value mod 3 (0, 1, 2) and then if you add the other value scaled by 3 you can have the msb doing x / 3 and lsb with x % 3 ? means the max value we store would be 6 which fits in 3 bits ?

Comment (tmontaigu) [reply tfhe/src/integer/server_key/comparator.rs:33]: We could, but that would not give room for improvements for 2_2 and we don't use parameters which have only 3 bits in integers, we jump from 2 to 4

Comment (IceTDrinker) [reply tfhe/src/integer/server_key/comparator.rs:33]: ok no worries

**Comment by tmontaigu** [comment]
No as I haven't yet asked for a review, but I'll do it after rebasing and handling the conflict

File: tfhe/src/integer/server_key/comparator.rs [hunk 2]
```diff
@@ -251,12 +232,61 @@ impl<'a> Comparator<'a> {
...
+
+        // We don't use pack_block_assign as the offset '4' does not depend on params
+        let mut result = self.server_key.key.unchecked_scalar_mul(&sign_blocks[1], 4);
+        self.server_key
+            .key
+            .unchecked_add_assign(&mut result, &sign_blocks[0]);
+        self.server_key
+            .key
+            .apply_lookup_table_assign(&mut result, &final_lut);
+        result
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/comparator.rs:274]: bivariate PBS does not fit here ?

File: tfhe/src/integer/server_key/comparator.rs [hunk 3]
```diff
@@ -251,12 +232,61 @@ impl<'a> Comparator<'a> {
...
+    fn reduce_signs<F>(
         &self,
         mut sign_blocks: Vec<crate::shortint::Ciphertext>,
-    ) -> crate::shortint::Ciphertext {
+        sign_result_handler_fn: F,
+    ) -> crate::shortint::Ciphertext
+    where
+        F: Fn(u64) -> u64,
+    {
+        while sign_blocks.len() != 2 {
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/comparator.rs:243]: should this be > 2 as in the function below ?

File: tfhe/src/integer/server_key/comparator.rs [hunk 4]
```diff
@@ -273,7 +303,30 @@ impl<'a> Comparator<'a> {
...
-        sign_blocks.into_iter().next().unwrap()
+
+        if sign_blocks.len() == 2 {
+            let final_lut = self.server_key.key.generate_lookup_table(|x| {
+                let final_sign = reduce_two_orderings_function(x);
+                sign_result_handler_fn(final_sign)
+            });
+            // We don't use pack_block_assign as the offset '4' does not depend on params
+            let mut result = self.server_key.key.unchecked_scalar_mul(&sign_blocks[1], 4);
+            self.server_key
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/comparator.rs:314]: bivariate PBS ?

File: tfhe/src/integer/server_key/comparator.rs [hunk 5]
```diff
@@ -562,104 +660,131 @@ impl<'a> Comparator<'a> {
...
-                            }
-                        });
-                        self.server_key.key.apply_lookup_table(
-                            &most_significant_blocks[most_significant_blocks.len() - 1],
-                            &lut2,
+
+                        let lut = self.server_key.key.generate_lookup_table_bivariate(
+                            |sign_block, msb_are_zeros| {
+                                let sign_bit_is_set = (sign_block >> sign_bit_pos) == 1;
+                                let sign_block_ordering = if sign_bit_is_set {
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/comparator.rs:688]: some comments could be welcome here for the if block, same a bit later for similar if blocks

**Review by IceTDrinker — COMMENTED** [review]
Some small things, notably maybe using the bivariate PBS directly in some places

could you comment on the change for the scalar code path ?

The main fix was applying the identity lut on packed blocks correct ? (any other fixes I have missed ?)

File: tfhe/src/integer/server_key/comparator.rs (hunk 2 above)
Comment (tmontaigu) [reply tfhe/src/integer/server_key/comparator.rs:274]: no it should

File: tfhe/src/integer/server_key/comparator.rs (hunk 3 above)
Comment (tmontaigu) [reply tfhe/src/integer/server_key/comparator.rs:243]: hum yes

**Review by IceTDrinker — APPROVED** [review]
Thanks !

**Pull request merged**